app/lib/actions/staypuft/host/puppet_run.rb

```diff
@@ -23,8 +23,23 @@ def plan(host)
         end
 
         def run
+          host_id = input.fetch(:host_id)
+
+          # try puppetrun 3 times
+          result = false
+          tries = 0
+          while !result && tries < 3
+            result = ::Host.find(host_id).puppetrun!
+            tries += 1
+          end
+
+          unless result
+            fail(::Staypuft::Exception, "Puppet run failed for host: #{host_id}")
```

Comment:
You might want to save the host.error_messages to get more details about the error into the action itself